docs/dev/testing.md

```diff
@@ -0,0 +1,188 @@
+
+In a unit test, we only want to test that this piece of code does its job.
+We don't care if any other piece of code does its job, and optimally a unit test should pass even if every other unit of code that it calls fails.
+
+To give an example from our current code base:
+
+```go
+func (c Client) StartBot(errChan chan<- error, sendMessage func(string)) {
+	fmt.Println("Starting up IRC bot...")
+	c.sendToTg = sendMessage
+	c.addHandlers()
+	if err := c.Connect(); err != nil {
+		errChan <- err
+	} else {
+		errChan <- nil
+	}
+}
+```
+
+In this example, we care that our function does the following things:
+
+* Passes "Starting up IRC bot..." to `fmt.Println`.
+* Sets `c.sendToTg` to the `sendMessage` passed in.
+* Calls `c.addHandlers`.
+* Calls `c.Connect`
+  * If an error is returned from this function, passed the error over the `errChan` channel passed in.
+  * Otherwise, passes `nil` over the `errChan` channel.
+
+Here are some things that we don't care about:
+
+* What `fmt.Println` does with our string, that's its responsibility.
+* What `c.addHandlers` does
+* What `c.Connect` does, we only want its return values
+
+In fact, we have an issue: if `c.Connect` _actually_ gets called, we can't guarantee that it will ever return what we want it to so we can actually test all of our cases.
+Not to mention, we want our unit tests to pass even if there is no network connection available.
+
+The solution to this: **mocking**.
+Due to static typing and the lack of inheritance in Go, the only real way to do proper mocking is to use interfaces.
+
+
+## Interfaces
+
+An interface in Go is similar to most other languages.
+It defines a contract that a concrete implementation must follow.
+
+For example:
+
+```go
+type Shape interface {
+    Area() float64
+}
+```
+
+This interface defines what a shape is: anything that has an area.
+A concrete implementation would be:
+
+```go
+type Rectangle struct {
+    Width  float64
+    Height float64
+}
+
+func (r Rectangle) Area() float64 {
+    return r.Width * r.Height
+}
+```
+
+Another implementation might be:
+
+```go
+type Circle struct {
+    Radius float64
+}
+
+func (c Circle) Area() float64 {
+    return math.Pi * math.Pow(c.Radius, 2)
+}
+```
+
+Now, we can create a function that accepts our interface:
+
+```go
+func PrintArea (s Shape) {
+    fmt.Println(s.Area())
+}
+
+func main() {
+	PrintArea(Rectangle{Width: 10, Height: 15}) // prints 150
+	PrintArea(Circle{Radius: 15}) // prints 706.8583
+}
+```
+
+### Why do we _need_ interfaces for testing?
+
+Because we cannot modify the implementation of functions on structs directly, we must change our functions to accept _interfaces_ instead.
+
+So, looking at one of our handlers:
+
+```go
+func connectHandler(c Client) func(*girc.Client, girc.Event) {
+	return func(gc *girc.Client, e girc.Event) {
+		c.Cmd.Join(c.Settings.Channel)
+	}
+}
+```
+
+In this case, you can see we have a bit of a problem: we're called `c.Cmd.Join`.
```

Comment:
"We're **calling** `c.Cmd.Join`"

Reply:
Fixed in commit 0d43c10cc8eff56728c9246a0c188c8120a2522e :+1: